### PR TITLE
shipit_uplift: Add tests for some coverage functions and the extensions API endpoint

### PR DESCRIPTION
**Pull request by gabriel-v**
closes https://github.com/mozilla-releng/services/issues/603

I probably should move more of those objects into json files under `fixtures/`.

### CODE REVIEW

**Comment by marco-c** [comment]
> I probably should move more of those objects into json files under fixtures/.

Yes, it would be more readable. Maybe split in different files (e.g. one per build), so when we add new ones in the future the file name can serve as documentation.

Since some endpoints require a rq worker to be running, the tests will probably timeout. We need to start a worker somehow, maybe in a test fixture with autouse=True.

File: src/shipit_uplift/tests/test_coverage.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    assert resp.status_code == 200
+    data = json.loads(resp.data.decode('utf-8'))
+    assert set(data) == set([
+        "c", "h", "cpp", "cc", "cxx", "hh", "hpp",
+        "hxx", "js", "jsm", "xul", "xml", "html", "xhtml",
+    ])
+
+
+
+def retry_request_for_result(client, url, retries=5, timeout=15):
```
Comment (marco-c) [inline src/shipit_uplift/tests/test_coverage.py]: You could use the `retry` function from cli_common.utils.

Comment (gabriel-v) [reply src/shipit_uplift/tests/test_coverage.py]: I only found a `retry` function under `shipit_code_coverage.utils`, and that's subjected to change. Don't see the `cli_common.utils` module in https://github.com/mozilla-releng/services/tree/master/lib/cli_common

Comment (marco-c) [reply src/shipit_uplift/tests/test_coverage.py]: Right, #849 was not merged yet :smile:

**Comment by gabriel-v** [comment]
I started the worker in an `autouse=True` fixture in the background with subprocess. The tests fail because there's no redis running on the Taskcluster testing environment.

So we either add redis to that environment, or add `fakeredis` as a dependency and mock the worker queue to run everything in the foreground.

Can't really find a way to install `fakeredis` on docker, as adding it in `requirements-dev.txt` makes nix complain about "attribute missing".

**Comment by marco-c** [comment]
A simpler solution might be to test the functions in the `coverage_by_changeset_impl`, `coverage_for_file_impl`, ... modules directly instead of the API endpoints.

Then in a separate PR, after we figure out the Redis story, we can add integration tests for the API endpoints. In this case we might want to use the real Redis, since they would be integration tests.

**Comment by marco-c** [comment]
@garbas, can you help regarding running Redis on the Taskcluster dev env? I guess we need to add support for starting it in https://github.com/mozilla-releng/services/blob/master/lib/please_cli/please_cli/run.py#L143 (but we don't want to run it on production, since on production we are not going to run Redis locally) and we need to install it in Nix (can we make it install only in the dev env?).

**Comment by marco-c** [comment]
Once https://github.com/mozilla-releng/services/pull/876 is merged, Redis should be available in the development environment.

@garbas suggested this nice package for running Redis in the context of the pytest tests: https://pypi.python.org/pypi/pytest-redis.

**Comment by gabriel-v** [comment]
I tested the API endpoints that didn't use rq and all the `*_impl.generate` functions.

Right now the tests hit the internet, but the results could be cached with some additional effort.

File: src/shipit_uplift/shipit_uplift/coverage.py [hunk 2]
```diff
@@ -260,14 +260,14 @@ def get_pushes_changesets(push_id, push_id_end):
...
-    if changeset not in changeset_cache:
+    if changeset[:12] not in changeset_cache:
         r = requests.get('https://hg.mozilla.org/mozilla-central/json-rev/%s' % changeset)
         rev = r.json()
         push_id = int(rev['pushid'])
 
         get_pushes(push_id)
 
-    return changeset_cache[changeset]
+    return changeset_cache[changeset[:12]]
```
Comment (gabriel-v) [inline src/shipit_uplift/shipit_uplift/coverage.py]: The changeset hash is truncated in [`get_pushes`](https://github.com/gabriel-v/services/blob/313d7d5bb716b08fed5c44d93a6e1d551d8e9c0c/src/shipit_uplift/shipit_uplift/coverage.py#L240) but not here

Comment (marco-c) [reply src/shipit_uplift/shipit_uplift/coverage.py]: Good catch. It was harmless because the changeset is truncated in the callers, but better to be safe than sorry.

File: src/shipit_uplift/tests/conftest.py [hunk 3]
```diff
@@ -118,3 +121,43 @@ def header_bot(app):
...
+    for path in sorted(paths):
+        with open(path) as f:
+            build_data = json.load(f)
+        builds['info'].update(build_data['info'])
+        builds['summary'].update(build_data['summary'])
+
+    return builds
+
+
+@pytest.yield_fixture(scope='session', autouse=True)
```
Comment (marco-c) [inline src/shipit_uplift/tests/conftest.py]: `yield_fixture` is deprecated, we should use the normal `fixture` which now supports yielding.

File: src/shipit_uplift/tests/conftest.py [hunk 4]
```diff
@@ -118,3 +121,43 @@ def header_bot(app):
...
+def run_worker():
+    '''
+    Run a worker in the background during the tests
+    '''
+    with subprocess.Popen(
```
Comment (marco-c) [inline src/shipit_uplift/tests/conftest.py]: Nit: put the following lines on one line

File: src/shipit_uplift/tests/test_coverage.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from __future__ import absolute_import
+
+import json
+import responses
+from shipit_uplift import \
```
Comment (marco-c) [inline src/shipit_uplift/tests/test_coverage.py]: Nit: use the same style as in api.py:
```Python
from shipit_uplift import (
    coverage_by_changeset_impl, coverage_summary_by_changeset_impl,
    coverage_for_file_impl
)
```

File: src/shipit_uplift/tests/conftest.py (hunk 4 above)
Comment (gabriel-v) [reply src/shipit_uplift/tests/conftest.py]: The fixture is actually not being used, as the tests don't hit rq anymore. Will remove it.

File: src/shipit_uplift/tests/test_coverage.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        coverage_by_changeset_impl, \
+        coverage_summary_by_changeset_impl, \
+        coverage_for_file_impl
+
+
+def test_coverage_supported_extensions_api(client):
```
Comment (marco-c) [inline src/shipit_uplift/tests/test_coverage.py]: It's fine for now, but in the future we might want to split the tests into multiple files (API endpoint tests in a file, coverage for changeset in another, and so on).

File: src/shipit_uplift/tests/test_coverage.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+def test_coverage_for_file_impl(client, coverage_changeset_by_file):
+    '''
+    Get code coverage information for a given file at a given changeset, from the internet
+    '''
+    responses.add_passthru('https://hg.mozilla.org/mozilla-central/')
+    responses.add_passthru('https://api.pub.build.mozilla.org/mapper/gecko-dev/rev')
+    responses.add_passthru('https://codecov.io/api/gh/marco-c/gecko-dev')
+    for file_coverage in coverage_changeset_by_file:
+        data = coverage_for_file_impl.generate(file_coverage['changeset'], file_coverage['path'])
+        assert data == file_coverage['data']
```
Comment (marco-c) [inline src/shipit_uplift/tests/test_coverage.py]: We can also add unit tests for functions in coverage.py, but we can do it in a separate PR.

File: src/shipit_uplift/tests/conftest.py (hunk 4 above)
Comment (marco-c) [reply src/shipit_uplift/tests/conftest.py]: Right, we can add it when writing the tests for the API endpoints which use rq.

File: src/shipit_uplift/tests/conftest.py [hunk 8]
```diff
@@ -11,9 +11,12 @@
 import os
 import pickle
 import pytest
+import subprocess
```
Comment (marco-c) [inline src/shipit_uplift/tests/conftest.py]: Need to remove unused imports and `WORKER_MODULE` too

File: src/shipit_uplift/tests/test_coverage.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from __future__ import absolute_import
+
+import json
+import responses
+from shipit_uplift import (
+        coverage_by_changeset_impl, coverage_summary_by_changeset_impl,
```
Comment (marco-c) [inline src/shipit_uplift/tests/test_coverage.py]: flake8 will probably complain about the 8 spaces here (or maybe not)

File: src/shipit_uplift/tests/conftest.py (hunk 8 above)
Comment (gabriel-v) [reply src/shipit_uplift/tests/conftest.py]: Thanks, I was waiting for taskcluster to explode for me.

**Comment by marco-c** [comment]
Would you mind filing issues for:
1) Adding tests for the coverage API endpoints that require Redis;
2) Adding unit tests for the functions in the coverage.py module.
